### PR TITLE
Add get group urdf capability

### PR DESCRIPTION
**Pull request by sjahr**
### Description

Requires: https://github.com/ros-planning/moveit_msgs/pull/174

This PR adds the move_group capability to request a urdf containing links and joints of a specific joint model group

### Testing
1. Launch move_group with the plugin loaded
2. Call service and request urdf of an existing group
```bash
ros2 service call /get_urdf moveit_msgs/srv/GetGroupUrdf group_name:\ \'MY_GROUP\'\
```

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit2/blob/main/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://moveit.picknik.ai/humble/doc/examples/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit2/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

**Comment by sjahr** [comment]
CI fails because https://github.com/ros-planning/moveit_msgs/pull/174 is not merged

File: moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+{
+}
+
+void GetUrdfService::initialize()
+{
+  get_urdf_service_ = context_->moveit_cpp_->getNode()->create_service<moveit_msgs::srv::GetGroupUrdf>(
+      GET_URDF_SERVICE_NAME,
+      [this](const std::shared_ptr<moveit_msgs::srv::GetGroupUrdf::Request>& req,
+             const std::shared_ptr<moveit_msgs::srv::GetGroupUrdf::Response>& res) {
+        auto const subgroup = context_->moveit_cpp_->getRobotModel()->getJointModelGroup(req->group_name);
```
Comment (marioprats) [inline moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp:68]: I think we are using 'const auto' in other places?

```suggestion
        const auto subgroup = context_->moveit_cpp_->getRobotModel()->getJointModelGroup(req->group_name);
```

File: moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        // Check if group exists in loaded robot model
+        if (!subgroup)
+        {
+          RCLCPP_ERROR(getLogger(), "Cannot create URDF because planning group %s does not exist",
+                       req->group_name.c_str());
+          res->success = false;
+          return;
+        }
+        // Get robot description string
+        auto full_urdf_string = std::string("");
```
Comment (marioprats) [inline moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp:78]:
```suggestion
        std::string full_urdf_string;
```

File: moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+          return;
+        }
+
+        // Create subgroup urdf
+        // Create header
+        res->urdf_string = std::string("<?xml version=\"1.0\" ?>\n<robot name=\"") + req->group_name +
+                           std::string("\" xmlns:xacro=\"http://ros.org/wiki/xacro\">");
+
+        // Create links
+        auto const link_names = subgroup->getLinkModelNames();
```
Comment (marioprats) [inline moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp:96]:
```suggestion
        const auto& link_names = subgroup->getLinkModelNames();
```

File: moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        for (const auto& link_name : link_names)
+        {
+          auto const start = full_urdf_string.find("<link name=\"" + link_name);
```
Comment (marioprats) [inline moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp:99]:
```suggestion
          const auto start = full_urdf_string.find("<link name=\"" + link_name);
```

File: moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+          auto substring = full_urdf_string.substr(start, full_urdf_string.size() - start);
+          res->urdf_string += substring.substr(0, substring.find(LINK_ELEMENT_CLOSING) + LINK_ELEMENT_CLOSING.size());
+        }
+        // Create joints
+        auto const joint_names = subgroup->getJointModelNames();
```
Comment (marioprats) [inline moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp:104]:
```suggestion
        const auto& joint_names = subgroup->getJointModelNames();
```

**Review by marioprats — COMMENTED** [review]
Thanks Sebastian!
Looks great! I have mostly comments about left/right const. Aren't we doing left const in other places? And I thought we had a lint check for it, but maybe not?

File: moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        {
+          const auto start = full_urdf_string.find("<link name=\"" + link_name);
+          auto substring = full_urdf_string.substr(start, full_urdf_string.size() - start);
+          res->urdf_string += substring.substr(0, substring.find(LINK_ELEMENT_CLOSING) + LINK_ELEMENT_CLOSING.size());
+        }
+        // Create joints
+        const auto& joint_names = subgroup->getJointModelNames();
+        for (const auto& joint_name : joint_names)
+        {
+          auto const start = full_urdf_string.find("<joint name=\"" + joint_name);
```
Comment (sjahr) [inline moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp:107]:
```suggestion
          const auto start = full_urdf_string.find("<joint name=\"" + joint_name);
```

**Review by marioprats — APPROVED** [review]
LGTM!

File: moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+#include <moveit/move_group/capability_names.h>
+#include <moveit/utils/logger.hpp>
+#include <tinyxml2.h>
+
+namespace move_group
+{
+
+namespace
+{
+const auto LOG_NAME = std::string("GetUrdfService");
```
Comment (henningkayser) [inline moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp:49]: Why the extra variable? Also, we seem to be mixing naming patterns for some reason, but the more consistent would be snake case

File: moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+          const auto error_string = std::string("Cannot create URDF because planning group '") + req->group_name +
+                                    std::string("' does not exist");
+          RCLCPP_ERROR(getLogger(), "%s", error_string.c_str());
+          res->error_code.message = error_string;
+          res->error_code.val = moveit_msgs::msg::MoveItErrorCodes::FAILURE;
+          return;
+        }
+        // Get robot description string
+        std::string full_urdf_string;
+        context_->moveit_cpp_->getNode()->get_parameter_or("robot_description", full_urdf_string, std::string(""));
```
Comment (henningkayser) [inline moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp:83]: Doesn't urdfdom help us with safely parsing the URDF from the RobotModel? It seems weird that we have utility libraries for managing URDF, but we have to build our custom parser.

**Review by henningkayser — APPROVED** [review]
Looks good in general. I still have some comments, but I wouldn't hold back the PR because of them.

File: moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+          const auto start = full_urdf_string.find("<joint name=\"" + joint_name);
+          auto substring = full_urdf_string.substr(start, full_urdf_string.size() - start);
+          res->urdf_string += substring.substr(0, substring.find(JOINT_ELEMENT_CLOSING) + JOINT_ELEMENT_CLOSING.size());
+        }
+        // Create closing
+        res->urdf_string += ROBOT_ELEMENT_CLOSING;
+
+        // Validate xml file
+        tinyxml2::XMLDocument group_urdf_xml;
+        group_urdf_xml.Parse(full_urdf_string.c_str());
```
Comment (sjahr) [inline moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp:123]: Use [parseURDF](https://github.com/ros/urdfdom/blob/29426ec119e9c3a9df9fc705d08c262a97b7f285/urdf_parser/src/check_urdf.cpp#L93) for validation

File: moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp (hunk 8 above)
Comment (sjahr) [reply moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp:83]: Good point, I can use [exportURDF](https://github.com/ros/urdfdom/blob/29426ec119e9c3a9df9fc705d08c262a97b7f285/urdf_parser/src/model.cpp#L312) and then use [accept](https://leethomason.github.io/tinyxml2/classtinyxml2_1_1_x_m_l_document.html#a853f05e2f7f51eb598e32ec61d8bc53c) to print the text. I will try this tomorrow morning

File: moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp (hunk 7 above)
Comment (sjahr) [inline moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp:49]:
```suggestion
```

File: moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+{
+}
+
+void GetUrdfService::initialize()
+{
+  get_urdf_service_ = context_->moveit_cpp_->getNode()->create_service<moveit_msgs::srv::GetGroupUrdf>(
+      GET_URDF_SERVICE_NAME,
+      [this](const std::shared_ptr<moveit_msgs::srv::GetGroupUrdf::Request>& req,
+             const std::shared_ptr<moveit_msgs::srv::GetGroupUrdf::Response>& res) {
+        res->error_code.source = LOG_NAME;
```
Comment (sjahr) [inline moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp:69]:
```suggestion
        res->error_code.source = std::string("GetUrdfService");
```

File: moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+
+namespace move_group
+{
+
+namespace
+{
+const auto LOG_NAME = std::string("GetUrdfService");
+rclcpp::Logger getLogger()
+{
+  return moveit::getLogger(LOG_NAME);
```
Comment (sjahr) [inline moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp:52]:
```suggestion
  return moveit::getLogger("get_urdf_service");
```

File: moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp (hunk 8 above)
Comment (sjahr) [reply moveit_ros/move_group/src/default_capabilities/get_group_urdf_capability.cpp:82]: Nope that did not work :disappointed: Keep using the parameter API for that

**Pull request merged**